File: apps/server/src/api-schemas.ts
import "./zod-openapi-init.js";
import { z } from "zod";

/**
 * Request-shape schemas for Mirror's native /api/* routes, shared between
 * index.ts (which validates against them at runtime) and
 * openapi-document.ts (which derives the generated OpenAPI document from
 * these same schemas). Splitting them out into their own module - rather
 * than defining them inline in index.ts, or letting openapi-document.ts
 * import them straight from index.ts - avoids a circular import between
 * index.ts and openapi-document.ts (index.ts already imports
 * buildOpenApiDocument from there to serve /mirror/openapi).
 */

export const SetSessionBody = z
  .object({
    sessionToken: z
      .string()
      .min(20, "That doesn't look like a valid session token")
      .openapi({
        description:
          "The value of the __Secure-next-auth.session-token cookie from chatgpt.com/api/auth/session.",
      }),
    turnstileToken: z
      .string()
      .optional()
      .openapi({
        description:
          "Optional Cloudflare Turnstile token for sentinel requirements.",
      }),
  })
  .openapi({ ref: "SetSessionBody" });

// Mirror conversation ids are UUIDs when auto-generated, but a caller can
// also name their own via /v1/chat/completions' metadata.conversation_id
// (e.g. an arbitrary slug) - accept any non-empty id here so a
// Playground/API-driven conversation created that way can still be
// browsed, loaded, and managed through these routes.
export const ConversationIdParam = z
  .object({
    id: z.string().min(1).openapi({ description: "A Mirror conversation id (UUID or a caller-supplied slug)." }),
  })
  .openapi({ ref: "ConversationIdParam" });

export const ModelUpdateBody = z
  .object({ model: z.string().min(1) })
  .openapi({ ref: "ModelUpdateBody", description: "Change a conversation's model." });

export const BranchBody = z
  .object({
    messageId: z.string().uuid().openapi({ description: "The message id to branch from; must have an upstreamNodeId." }),
    title: z.string().optional(),
  })
  .openapi({ ref: "BranchBody" });

export const NewConversationBody = z
  .object({
    model: z.string().default("auto"),
    gizmoId: z.string().nullable().optional(),
  })
  .openapi({ ref: "NewConversationBody" });

const queryBoolean = z.enum(["true", "false"]).transform((value) => value === "true");
export const ConversationsQuery = z
  .object({
    limit: z.coerce.number().int().min(1).max(200).default(50),
    offset: z.coerce.number().int().min(0).default(0),
    sync: queryBoolean.default("true").openapi({
      description: "Whether to sync this page against the real ChatGPT sidebar before reading the local mirror of it.",
      type: "string",
    }),
    // Restart the incremental sync cursor from the very top instead of
    // resuming where a previous call left off - only worth paying for on an
    // explicit user-initiated refresh (see App.tsx), not on every page load.
    resync: queryBoolean.default("false").openapi({
      description: "Restart the incremental sync cursor from the top instead of resuming where a previous call left off.",
      type: "string",
    }),
  })
  .openapi({ ref: "ConversationsQuery" });

export const AssetsQuery = z
  .object({
    pointer: z.string().openapi({ description: "A file-service:// or sediment:// asset pointer." }),
    upstreamConversationId: z.string().optional().openapi({
      description: "Required when pointer is a sediment:// pointer, to verify ownership of the source conversation.",
    }),
  })
  .openapi({ ref: "AssetsQuery" });


export const ChatAttachment = z
  .object({
    fileId: z.string(),
    fileName: z.string(),
    fileSize: z.number(),
    mimeType: z.string(),
    useCase: z.enum(["multimodal", "my_files"]),
    width: z.number().optional(),
    height: z.number().optional(),
    raw: z.record(z.unknown()).default({}),
  })
  .openapi({ ref: "ChatAttachment", description: "A file previously uploaded via POST /api/files." });

export const ChatBody = z
  .object({
    prompt: z.string().min(1),
    model: z.string().default("auto"),
    conversationId: z.string().uuid().nullable().optional(),
    gizmoId: z.string().nullable().optional(),
    timezone: z.string().optional(),
    timezoneOffsetMin: z.number().optional(),
    attachments: z.array(ChatAttachment).default([]),
    turnstileToken: z.string().optional().openapi({
      description: "Optional Cloudflare Turnstile token override for sentinel requirements.",
    }),
  })
  .openapi({
    ref: "ChatBody",
    description: "Mirror's native chat request, used by the proxied ChatGPT UI and Playground (see POST /api/chat).",
  });

File: crates/mirror-server/src/api_schemas.rs
//! Request and response schemas for Mirror's native `/api/*` and `/v1/*` routes.
//! Port of `apps/server/src/api-schemas.ts`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SetSessionBody {
    #[serde(rename = "sessionToken")]
    pub session_token: String,
    #[serde(rename = "turnstileToken", skip_serializing_if = "Option::is_none")]
    pub turnstile_token: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ConversationIdParam {
    pub id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ModelUpdateBody {
    pub model: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BranchBody {
    #[serde(rename = "messageId")]
    pub message_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
}

fn default_auto_model() -> String {
    "auto".to_string()
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NewConversationBody {
    #[serde(default = "default_auto_model")]
    pub model: String,
    #[serde(rename = "gizmoId", skip_serializing_if = "Option::is_none")]
    pub gizmo_id: Option<String>,
}

fn default_limit() -> usize {
    50
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ConversationsQuery {
    #[serde(default = "default_limit")]
    pub limit: usize,
    #[serde(default)]
    pub offset: usize,
    #[serde(default = "default_true")]
    pub sync: bool,
    #[serde(default)]
    pub resync: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AssetsQuery {
    pub pointer: String,
    #[serde(rename = "upstreamConversationId", skip_serializing_if = "Option::is_none")]
    pub upstream_conversation_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ChatAttachment {
    #[serde(rename = "fileId")]
    pub file_id: String,
    #[serde(rename = "fileName")]
    pub file_name: String,
    #[serde(rename = "fileSize")]
    pub file_size: u64,
    #[serde(rename = "mimeType")]
    pub mime_type: String,
    #[serde(rename = "useCase")]
    pub use_case: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<u32>,
    #[serde(default)]
    pub raw: Value,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ChatBody {
    pub prompt: String,
    #[serde(default = "default_auto_model")]
    pub model: String,
    #[serde(rename = "conversationId", skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
    #[serde(rename = "gizmoId", skip_serializing_if = "Option::is_none")]
    pub gizmo_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timezone: Option<String>,
    #[serde(rename = "timezoneOffsetMin", skip_serializing_if = "Option::is_none")]
    pub timezone_offset_min: Option<i32>,
    #[serde(default)]
    pub attachments: Vec<ChatAttachment>,
    #[serde(rename = "turnstileToken", skip_serializing_if = "Option::is_none")]
    pub turnstile_token: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn chat_body_deserializes_with_defaults() {
        let raw = json!({ "prompt": "hello world" });
        let body: ChatBody = serde_json::from_value(raw).unwrap();
        assert_eq!(body.prompt, "hello world");
        assert_eq!(body.model, "auto");
        assert!(body.attachments.is_empty());
        assert_eq!(body.conversation_id, None);
    }

    #[test]
    fn set_session_body_round_trips() {
        let raw = json!({
            "sessionToken": "test-session-token-12345",
            "turnstileToken": "cf-token"
        });
        let body: SetSessionBody = serde_json::from_value(raw).unwrap();
        assert_eq!(body.session_token, "test-session-token-12345");
        assert_eq!(body.turnstile_token.as_deref(), Some("cf-token"));
    }

    #[test]
    fn conversation_param_and_model_update_round_trip() {
        let id_raw = json!({ "id": "conv-123" });
        let id_param: ConversationIdParam = serde_json::from_value(id_raw).unwrap();
        assert_eq!(id_param.id, "conv-123");

        let model_raw = json!({ "model": "gpt-4o" });
        let model_body: ModelUpdateBody = serde_json::from_value(model_raw).unwrap();
        assert_eq!(model_body.model, "gpt-4o");
    }

    #[test]
    fn branch_and_new_conversation_schemas() {
        let branch_raw = json!({
            "messageId": "msg-123",
            "title": "Branch Title"
        });
        let branch: BranchBody = serde_json::from_value(branch_raw).unwrap();
        assert_eq!(branch.message_id, "msg-123");
        assert_eq!(branch.title.as_deref(), Some("Branch Title"));

        let new_conv_raw = json!({ "gizmoId": "g-abc" });
        let new_conv: NewConversationBody = serde_json::from_value(new_conv_raw).unwrap();
        assert_eq!(new_conv.model, "auto");
        assert_eq!(new_conv.gizmo_id.as_deref(), Some("g-abc"));
    }

    #[test]
    fn conversations_and_assets_query_schemas() {
        let query_raw = json!({
            "limit": 25,
            "offset": 50,
            "sync": false,
            "resync": true
        });
        let query: ConversationsQuery = serde_json::from_value(query_raw).unwrap();
        assert_eq!(query.limit, 25);
        assert_eq!(query.offset, 50);
        assert!(!query.sync);
        assert!(query.resync);

        let asset_raw = json!({
            "pointer": "file-service://file-1",
            "upstreamConversationId": "up-conv-2"
        });
        let asset: AssetsQuery = serde_json::from_value(asset_raw).unwrap();
        assert_eq!(asset.pointer, "file-service://file-1");
        assert_eq!(asset.upstream_conversation_id.as_deref(), Some("up-conv-2"));
    }

    #[test]
    fn chat_attachment_schema() {
        let attach_raw = json!({
            "fileId": "f-123",
            "fileName": "photo.png",
            "fileSize": 1024,
            "mimeType": "image/png",
            "useCase": "multimodal",
            "width": 100,
            "height": 200
        });
        let attach: ChatAttachment = serde_json::from_value(attach_raw).unwrap();
        assert_eq!(attach.file_id, "f-123");
        assert_eq!(attach.file_name, "photo.png");
        assert_eq!(attach.file_size, 1024);
        assert_eq!(attach.mime_type, "image/png");
        assert_eq!(attach.use_case, "multimodal");
        assert_eq!(attach.width, Some(100));
        assert_eq!(attach.height, Some(200));
    }
}
